src/core/components/InstallButton/index.js

```diff
@@ -71,6 +70,18 @@ export class InstallButtonBase extends React.Component {
     return i18n.sprintf(label, { name });
   }
 
+  getDownloadProgress() {
```

Comment:
1. `InstallButton` is a wrapper around `Switch` that provides the hooks for installing an add-on or theme and tells the `Switch` what state to be in. This logic is how an add-on install should be displayed by a switch. Likewise if we had a switch that saved a user setting it might only do the `Infinity` and `-Infinity` case if it saves on toggle.
2. I think fetching data from the redux store at the lowest level can only limit our options in how something is used. By creating components that don't interact with redux at the lowest level we should be able to wrap them in very basic containers that just provide some logic like this one. That way the code for creating this pretty switch that we might use in multiple places isn't tied to downloads.

> The key difference in this model is that the entire chain of components is only re-rendered once when the downloadTransactionId changes, not all the time as downloadProgress gets updated repeatedly.

Do you mean this prevents the `Addon` component from re-rendering? The `Switch` will always need to re-render so there's no benefit there. This sounds like an optimisation on something that's already optimised (React virtual DOM diffing).